File: data_structure/js/LinkedList/Linked-List.js
class Node {
	constructor(value) {
		this.value = value
		this.next = null
	}
}

class LL {
	constructor() {
		this.head = null
	}

	pushFront(value) {
		let node = new Node(value)
		if (this.head == null) {
			this.head = node
		} else {
			node.next = this.head
			this.head = node			
		}
	}

	pushRear(value) {
		let node = new Node(value)
		if (this.head == null) {
			this.head = node
		} else {
			let tmp = this.head
			while (tmp.next != null) {
				tmp = tmp.next
			}
			tmp.next = node
		}	
	}

	delete(value) {
		if (this.head == null) return
		else {
			if (this.head.value == value) {
				this.head = this.head.next
			} else {
				let tmp = this.head
				while (tmp.next != null) {
					if (tmp.next.value == value) {
						tmp.next = tmp.next.next
						break
					}
					tmp = tmp.next
				}
			}
		}
	}

	traversal() {
		let ans = ''
		if (this.head != null) {
			let tmp = this.head
			while (tmp != null) {
				ans += tmp.value.toString() + ' '
				tmp = tmp.next
			}
		}
		console.log(ans)
	}

	reverse() {
		if (this.head == null) return
		else {
			let previous = null
			let current = this.head
			let cNext = this.head.next
			while (cNext != null) {
				current.next = previous
				previous = current
				current = cNext
				cNext = cNext.next
			}
			current.next = previous
			this.head = current
		}
	}
}

let list = new LL()
list.pushFront(4)
list.pushFront(6)
list.pushFront(2)
list.pushRear(3)
list.pushRear(9)
list.pushFront(1)
list.pushRear(8)
list.delete(1)
list.delete(8)
list.delete(4)
list.traversal()
list.reverse()
list.traversal()

File: grad/Linked_List.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node {
	int value;
	struct node *next;
} Node;

Node *head;

void insert(int value) {
	
	Node *node = NULL;
	node = (Node*)malloc(sizeof(Node));

	node->value = value;
	node->next = NULL;

	if (head == NULL) {
		head = node;
		return;
	}

	Node *tmp = head;
	while (tmp->next != NULL)
		tmp = tmp->next;
	tmp->next = node;
}

void traversal(Node *current) {
	if (current == NULL) return;
	while (current) {
		printf("%d -> ", current->value);
		current = current->next;
	}
	printf("NULL\n");
}

void delete(int target) {
	if (head == NULL) return;
	
	if (head->value == target) {
		head = head->next;
		return;
	}

	Node *current = head;
	while (current->next) {
		if (current->next->value == target) {
			current->next = current->next->next;
			return;
		}
		current = current->next;
	}
}
void reverse() {
	if (head == NULL) return;

	Node *prev = NULL;
	Node *curr = NULL;
	Node *next = head;
	printf("hhh\n");
	while (next) {
		prev = curr;
		curr = next;
		next = next->next;
		curr->next = prev;
	}

	head = curr;
}
int main() {
	int mode = 0;
	int value = -1;
	while (1) {
		
		printf("0)exit 1)insert 2)delete 3)reverse ... ");
		fflush(stdin);
		scanf("%d", &mode);
		
		if (mode == 0) {
			exit(0);
		} else if (mode == 1) {
			scanf("%d", &value);
			insert(value);
		} else if (mode == 2) {
			scanf("%d", &value);
			delete(value);
		} else if (mode == 3) {
			reverse();
		}
		traversal(head);
	}
}